src/core/unit.c

```diff
@@ -3180,6 +3180,54 @@ static int resolve_template(Unit *u, const char *name, char **buf, const char **
         return 0;
 }
 
+static int unit_is_recursive_template_dependency(Unit *u, const char *name) {
+        _cleanup_free_ char *instance = NULL;
+        const char *fragment_path;
+        _cleanup_set_free_ Set *names = NULL;
+        int r;
+
+        assert(u);
+        assert(name);
+
+        /* If a template unit has a direct dependency on itself that contains the unit name as part of the
+         * template instance, this will almost certainly lead to infinite recursion as systemd will keep
+         * instantiating new instances of the template unit. https://github.com/systemd/systemd/issues/17602
+         * shows a good example of how this can happen in practice. To guard against this, we check for
+         * templates that depend on themselves and have the instantiated unit name as part of the template
+         * instance of the dependency.
+         */
+
+        if (!unit_name_prefix_equal(u->id, name))
+                return 0;
+
+        if (u->type != unit_name_to_type(name))
+                return 0;
+
+        r = unit_file_find_fragment(u->manager->unit_id_map, u->manager->unit_name_map, name, &fragment_path, &names);
+        if (r < 0)
+                return r;
+
+        if (!u->fragment_path || !fragment_path)
+                return 0;
+
+        /* Fragment paths should also be equal as a custom fragment for a specific template instance
+         * wouldn't necessarily lead to infinite recursion. */
+        if (!streq(u->fragment_path, fragment_path))
+                return 0;
+
+        r = unit_name_to_instance(name, &instance);
+        if (r < 0)
+                return r;
+
+        if (!u->instance || !instance)
+                return 0;
+
+        if (!strstr(instance, u->instance))
```

Comment:
hmm, this is black magic. why not streq()?

Reply:
In the example in the linked issue, the global drop-in looks as follows:

```
[Unit]
Wants=notify@%n.service
```

Here the instance will be the full unit name, so `streq` would fail, but this still leads to infinite recursion as the full unit name contains the instance. That's why we use `strstr`, to catch any string that contains the instance, since all of those will almost surely lead to infinite recursion (If only the instance is sufficient to get recursion, every string containing the instance will surely do so as well).